### DISCUSSION TITLE
[General]: v8 to v9 Migration Guide (breaking changes)

### DISCUSSION THREAD

**Discussion by davidjerleke** [davidjerleke/embla-carousel#1080] [General]
## Migration Guide (Breaking Changes Only)

This guide lists only breaking changes introduced in v9.0.0 and shows how to achieve the same behavior as in v8.6.0.

---

### 1.1 Method Renames (Breaking)

| v8 | v9 |
|---|---|
| `scrollNext()` | `goToNext()` |
| `scrollPrev()` | `goToPrev()` |
| `canScrollNext()` | `canGoToNext()` |
| `canScrollPrev()` | `canGoToPrev()` |
| `scrollTo(index, jump?, direction?)` | `goTo(index, instant?, direction?)` |
| `scrollSnapList()` | `snapList()` |
| `selectedScrollSnap()` | `selectedSnap()` |
| `previousScrollSnap()` | `previousSnap()` |
| `slidesNotInView()` | ❌ Removed |

### 1.2 Updated Usage Examples

```ts
emblaApi.goToNext(/* instant?: boolean */)
emblaApi.goToPrev(/* instant?: boolean */)

emblaApi.goTo(targetIndex, /* instant?: boolean */, /* 'forward' | 'backward' */)

emblaApi.snapList()
emblaApi.selectedSnap()
emblaApi.previousSnap()
```

### 1.3 Replacement for slidesNotInView()

```ts
const inView = emblaApi.slidesInView()
const all = emblaApi.slideNodes().map((_, i) => i)
const notInView = all.filter((i) => !inView.includes(i))
```

Note: The instant-scroll flag remains a boolean (jump → instant).

### 2.1 General Event Changes (Breaking)

- Event names are lowercase.
- The `init` event has been removed.
- Event callbacks now receive `(api, event)` with `event.type` and `event.detail`.

```ts
(api, event) => {
  event.type
  event.detail
}
```

### 2.2 v8 → v9 Event Mapping

| v8 | v9 | Notes |
|---|---|---|
| `pointerDown` | `pointerdown` | Case change |
| `pointerMove` | `pointermove` | Case change |
| `pointerUp` | `pointerup` | Case change |
| `slidesInView` | `slidesinview` | Case change |
| `slidesChanged` | `slideschanged` | Case change |
| `slideFocusStart` | `slidefocus` | Renamed |
| `scroll` | `scroll` | Unchanged |
| `select` | `select` | Unchanged |
| `settle` | `settle` | Unchanged |
| `resize` | `resize` | Detail type changed |
| `reInit` | `reinit` | Case change |
| `destroy` | `destroy` | Unchanged |
| `init` | ❌ removed | Run setup after init |

### 2.3 Full Event List (v9)

Events marked NEW did not exist in v8.

```ts
export interface EmblaEventListType {
  pointerdown: PointerEventType
  pointermove: PointerEventType
  pointerup: PointerEventType
  slideschanged: MutationRecord[]
  slidesinview: SlidesInViewEventType
  scrolloptimize: ScrollOptimizeEventType // NEW
  select: SelectEventType
  scroll: ScrollEventType
  settle: null
  destroy: null
  reinit: null
  resize: ResizeObserverEntry[]
  slidefocus: FocusEvent
}
```

### 2.4 Event Usage Examples

```ts
emblaApi.on('pointerdown', (api, event) => {})
emblaApi.on('slidesinview', (api, event) => {})
emblaApi.on('slidefocus', (api, event) => {})
emblaApi.on('reinit', () => {})
```

### 2.5 No init Event Replacement

**v8**

```ts
const emblaApi = EmblaCarousel(root, options, plugins)

function mySetupFunction() {
  // run setup logic here
}

emblaApi.on('init', mySetupFunction)
```

**v9**

```ts
const emblaApi = EmblaCarousel(root, options, plugins)

function mySetupFunction() {
  // run setup logic here
}

mySetupFunction()
```

### 2.6 Cancellable Events

- Returning `false` from a handler cancels the event's internal handling.
- The former `watchDrag` callback maps to the cancellable `pointerdown` event.

```ts
emblaApi.on('pointerdown', (api, event) => {
  return false // cancel pointer down internal logic
})

emblaApi.on('pointerdown', (api, event) => {
  allow pointer down internal logic
  return true // allow pointer down internal logic
})

emblaApi.on('pointerdown', (api, event) => {
  // allow pointer down internal logic
})
```

## 3. Options (Breaking Renames)

| v8 option | v9 option | Notes |
|---|---|---|
| `startIndex` | `startSnap` | Selected snap index at init. |
| `watchDrag` | `draggable` | Enables/disables drag interactions. v8 callback moved to cancellable `pointerdown` event. |
| `watchResize` | `resize` | Enables/disables resize observing. v8 callback moved to cancellable `resize` event. |
| `watchSlides` | `slideChanges` | Enables/disables slide mutation observing. Related event `slideschanged` is cancellable. |
| `watchFocus` | `focus` | Enables/disables focus handling. Related event `slidefocus` is cancellable. |

Example option mapping:

```ts
// v8
const optionsV8 = {
  startIndex: 0,
  watchDrag: true,
  watchResize: true,
  watchSlides: true,
  watchFocus: true
}

// v9
const optionsV9 = {
  startSnap: 0,
  draggable: true,
  resize: true,
  slideChanges: true,
  focus: true
}
```

Callback migration:

| v8 callback option | v9 event | How to cancel |
|---|---|---|
| `watchDrag` (fn) | `pointerdown` | Return `false` from handler to cancel pointer move/drag handling. |
| `watchResize` (fn) | `resize` | Return `false` from handler to cancel Embla's internal resize handling. |
| `watchFocus` (fn) | `slidefocus` | Return `false` from handler to cancel Embla's internal resize handling. |
| `watchSlides` (fn) | `slideschanged` | Return `false` from handler to cancel Embla's internal resize handling. |

### 4.1 Autoplay (embla-carousel-autoplay)

Breaking option changes:

- Removed: `playOnInit`, `stopOnFocusIn`, `stopOnInteraction`, `stopOnMouseEnter`
- Renamed: `jump` → `instant`
- Added: `defaultInteraction: boolean`

```ts
import Autoplay from 'embla-carousel-autoplay'

const autoplay = Autoplay({ instant: false })
const emblaApi = EmblaCarousel(root, options, [autoplay])

// Since playOnInit has been removed, manually start it like so:
emblaApi.plugins().autoplay.play()

// Custom interaction handling instead of the now removed `stopOnFocusIn`, `stopOnInteraction`, `stopOnMouseEnter`:
const autoplay = Autoplay({ defaultInteraction: false })

emblaApi.on('autoplay:interaction', (api, event) => {
  const { interaction, originalEvent, isMouseOver, isPointerDown } = event.detail
})

// Default interaction will stop on pointerdown and slidefocus
```

### 4.2 Auto Scroll (embla-carousel-auto-scroll)

Breaking option changes:

- Removed: `playOnInit`, `stopOnFocusIn`, `stopOnInteraction`, `stopOnMouseEnter`
- Added: `defaultInteraction: boolean`

```ts
import AutoScroll from 'embla-carousel-auto-scroll'

const autoScroll = AutoScroll({ defaultInteraction: true })
const emblaApi = EmblaCarousel(root, options, [autoScroll])

// Since playOnInit has been removed, manually start it like so:
emblaApi.plugins().autoScroll.play()

// Custom interaction handling instead of the now removed `stopOnFocusIn`, `stopOnInteraction`, `stopOnMouseEnter`:
const autoScroll = AutoScroll({ defaultInteraction: false })

emblaApi.on('autoscroll:interaction', (api, event) => {
  const { interaction, originalEvent, isMouseOver, isPointerDown } = event.detail
})

// Default interaction will stop on pointerdown and slidefocus
```

### 4.3 Auto Height (embla-carousel-auto-height)

Breaking change:

- Height is now applied to the root node (previously the container).
- Added option: `heightEvent: 'select' | 'slidesinview'` (default: `select`).

```ts
import AutoHeight from 'embla-carousel-auto-height'

const autoHeight = AutoHeight({ heightEvent: 'select' })
const emblaApi = EmblaCarousel(root, options, [autoHeight])

emblaApi.rootNode().style.height
```

### 4.4 Class Names (embla-carousel-class-names)

| v8 | v9 |
|---|---|
| `dragging` | `pointerDown` |

```ts
import ClassNames from 'embla-carousel-class-names'

ClassNames({ pointerDown: 'is-pointer-down' })
```

### Vue (Breaking)

```ts
import useEmblaCarousel from 'embla-carousel-vue'

const [rootNode, emblaApiRef, emblaSsrApi] = useEmblaCarousel(options, plugins)
```

### Svelte (Breaking)

```svelte
<div use:useEmblaCarousel={{ options, plugins }} on:emblainit={(e) => console.log(e.detail)}>
  <!-- slides -->
</div>
```

### Solid (Breaking)

`createEmblaCarousel` → `useEmblaCarousel`

### 6.1 Event Emission (Breaking)

```ts
// v8
emblaApi.emit('autoplay:play')

// v9
emblaApi.createEvent('autoplay:play', null).emit()
```

### 6.2 Internal API Changes (Breaking)

| v8 | v9 |
|---|---|
| `internalEngine().ownerWindow` | `internalEngine().nodeHandler.ownerWindow` |
| `internalEngine().ownerDocument` | `internalEngine().nodeHandler.ownerDocument` |
| `index` | `indexCurrent` |
| `axis.measureSize` | `axis.getSize` |
| `limit.reachedMin` | `pastMinBound` |
| `limit.reachedMax` | `pastMaxBound` |
| `limit.constrain` | `clamp` |

## 7. Version References

- v8 baseline commit: `0848932`
- v9: `HEAD`

If additional breaking changes are found, please open an issue with code references and this guide will be updated.

**Comment by ttnbtfy** [comment]
Hello, I have some `init` events like this:
```js
this.slider.on( 'init', () => this.classList.add( 'hello-world' ));
```

How can I change the structure in version v9?

**Reply by davidjerleke** [reply]
@ttnbtfy thanks for the question, I've updated the migration guide with a clear example now. Please see the **2.5 No init Event Replacement** section.

**Comment by sarussss** [comment]
Hi @davidjerleke 

Thank you for your efforts in creating Embla v9.
I have two questions: 
Will the package size increase significantly with SSR support? This is a major concern for those who don't use it, like me.
My second question is, with the current V9 APIs, can we create [card effects ](https://swiperjs.com/demos#effect-cards)like this in Embla?

**Reply by davidjerleke** [reply]
Hi @sarussss,

Thanks for the kind words! I’m glad you’re interested in Embla v9.

Just a heads-up: v9 is currently in **RC** (release candidate), so some things could still change, like bundle size improvements.

Regarding **SSR support and bundle size**, I know many developers appreciate Embla’s slim size, and I’ll do my best to avoid adding much overhead. For example, I’m considering whether some features could be moved out of the core and offered as plugins instead. That said, Embla isn’t primarily aimed at developers who are solely focused on minimizing KB — it’s not the absolute smallest carousel library. Rather, it’s a **lightweight, extensible library** that prioritizes a flexible API. There are smaller carousel libs that cover very simple use cases, but developers often run into limitations as their needs grow.

As for **card effects**, it’s definitely possible to build them on top of Embla, similar to iOS pickers. For reference, see [this example](https://www.embla-carousel.com/examples/predefined/#ios-style-picker-default). This was also possible in v8 with `emblaApi.internalEngine()`, but I haven’t had the time to create a dedicated example yet, since the usage of this kind of effect is more limited than the core features.

Hope this helps clarify things!

**Reply by sarussss** [reply]
@davidjerleke Thank you for your answer, it helped me a lot.

**Reply by davidjerleke** [reply]
Hi @sarussss and @patricekaufmann,

Just a quick update for you and anyone interested: I’m working on moving the SSR feature into a dedicated plugin to keep the core lean and reduce the `v9` bundle size:

- #1326

**Reply by davidjerleke** [reply]
Hi @sarussss,

I just released [9.0.0-rc02](https://www.npmjs.com/package/embla-carousel/v/9.0.0-rc02) where SSR is a dedicated plugin and not part of the core. Enjoy!

**Comment by MartinCura** [comment]
Migrated v8 -> v9 RC.1 without any trouble, seems to be working great :)

**Reply by davidjerleke** [reply]
Thanks for letting me now @MartinCura 👍. 

I just released [9.0.0-rc02](https://www.npmjs.com/package/embla-carousel/v/9.0.0-rc02) which includes important bug fixes. Feel free to upgrade and try it out.

**Reply by MartinCura** [reply]
Didn't face those bugs but i'll surely upgrade at some point, thank you for your work
